wingtips-core/src/main/java/com/nike/wingtips/Span.java

```diff
@@ -357,19 +360,38 @@ public SpanPurpose getSpanPurpose() {
     }
 
     /**
-     * Indicates that this {@link Span} is completed/finished/finalized and sets {@link #getDurationNanos()} to be {@link System#nanoTime()} minus
-     * {@link #getSpanStartTimeNanos()}. After this is called then {@link #isCompleted()} will return true and {@link #getDurationNanos()} will return
-     * the value calculated here. An {@link IllegalStateException} will be thrown if this method is called after the span has already been completed.
-     * <p/>
-     * NOTE: This is intentionally package scoped to make sure completions and logging/span output logic happens centrally through {@link Tracer}.
+     * Causes this {@link Span} to be completed/finished/finalized by setting {@link #getDurationNanos()} to
+     * {@link System#nanoTime()} minus {@link #getSpanStartTimeNanos()}. After this is called then
+     * {@link #isCompleted()} will return true and {@link #getDurationNanos()} will return the value calculated here.
+     * A return value of true means this span wasn't previously completed, and this method call was the one that
+     * caused it to be completed. This is done atomically to guarantee only one caller can ever return true for this
+     * method for a given span. If this span has already been completed, then this method will return false and won't
+     * do anything.
+     *
+     * <p>This atomic completion behavior is necessary to support some asynchronous use cases, where multiple threads
+     * may try to complete the span at the same time (most often in the case of a problem that can trigger multiple
+     * simultaneous errors).
+     *
+     * <p>NOTE: This is intentionally package scoped to make sure completions and logging/span output logic happens
+     * centrally through {@link Tracer}.
+     *
+     * @return true if this span was *not* previously completed and the call to this method caused it to become
+     * completed, false if this span was previously completed (and therefore the call to this method did nothing).
      */
-    /*package*/ void complete() {
-        if (this.durationNanos != null)
-            throw new IllegalStateException("This Span is already completed.");
+    /*package*/ boolean complete() {
```

Comment:
does the return value result in anything special? otherwise I would just not return it and instead offer a hook, metric or log for redundantly closed spans.

iotw I doubt you would want each instrumentation to check this as they may just be unlucky due to a buggy other..

Reply:
App-level instrumentation won't ever deal with this. The boolean return value is used by `Tracer` to decide whether it should continue on its span-completion duties, i.e. should `Tracer` trigger `spanCompleted(Span)` on all the `SpanLifecycleListener`s registered to receive span-completion events, and should `Tracer` log the completed span? We only want those things to happen once.